### ISSUE TITLE
[MNG-5626] Avoid negative durations or handle them correctly

### ISSUE THREAD

**Issue by jira-importer** [apache/maven#6693]
**[Christian Jung](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=christian.jung)** opened **[MNG-5626](https://issues.apache.org/jira/browse/MNG-5626?redirect=false)** and commented

In issue MNG-5623 we reported an exception when printing the reactor summary if one of the times was negative.

I saw in one case, that the overall maven build time, as measured from outside (i.e. by our QuickBuild system) was -10.8 seconds. The corresponding reactor summary was:

```
13:55:25,184 INFO  - Reactor Summary:
13:55:25,184 INFO  -
13:55:25,184 INFO  - module1 ........................................... SUCCESS [ 5.911 s]
13:55:25,184 INFO  - module2 ........................................... SUCCESS [ 0.255 s]
13:55:25,184 INFO  - gpPlaygroundBase-lnx-x64-gcc4 ..................... SUCCESS [-27.-64 s]
13:55:25,185 INFO  - ------------------------------------------------------------------------
13:55:25,185 INFO  - BUILD SUCCESS
13:55:25,185 INFO  - ------------------------------------------------------------------------
13:55:25,185 INFO  - Total time: -20.-73 s
13:55:25,185 INFO  - Finished at: 2014-04-28T13:55:25+01:00
13:55:25,572 INFO  - Final Memory: 32M/439M
13:55:25,572 INFO  - ------------------------------------------------------------------------
```

The thing is quite hard to reproduce, the machines were virtual machines that have been running for quite a long time.

Our administrators suspected that just at this point, the local clock was synchronized with some outer source.

We should check if such negative durations can be avoided, and if not, they should be handled correctly.

---

**Affects:** 3.2.1

**Issue Links:**
- [MNG-5623](https://issues.apache.org/jira/browse/MNG-5623) Exception when printing Reactor Summary

- [MNG-5623](https://issues.apache.org/jira/browse/MNG-5623) Exception when printing Reactor Summary

- [MNG-5729](https://issues.apache.org/jira/browse/MNG-5729) Measure (build time) durations with System.nanoTime only
 (_**"depends upon"**_)

**Backported to:** [4.0.0-rc-2](https://github.com/apache/maven/milestone/27?closed=1)

0 votes, 9 watchers

**Comment by jira-importer** [comment]
**[Christian Jung](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=christian.jung)** commented

I found a simpler way to reproduce the error: With the following simple script i reset the time every few seconds:
{code:timewarp.sh}
#!/bin/bash
while true; do
date +%T -s "10:13:13"
sleep 5
date
done

```
(This is linux of course) This produces negative durations in maven. The output of `time maven ...` is then:
```

\<snip/>
INFO] ------------------------------------------------------------------------
[INFO] Reactor Summary:
[INFO] 
[INFO] module1 ........................................... SUCCESS [  0.002 s]
[INFO] module2 ........................................... SUCCESS [  0.000 s]
[INFO] gpPlaygroundBase-lnx-x86-gcc4 ..................... SUCCESS [-2.-497 s]
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time: -1.-774 s
[INFO] Finished at: 2014-04-29T10:13:14+01:00
[INFO] Final Memory: 17M/310M
[INFO] ------------------------------------------------------------------------

real    0m0.000s
user    0m6.148s
sys     0m0.996s

```
It seems that `time` does not display any negative times.
```

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

The bash snippet requires root permission for the date. Now we need to locate the cause, the build summary class or the JVM.
Can you write a simple class which calls `currentTimeMillis` with an arbitrary pause twice and check whether you see a negative difference?

**Comment by jira-importer** [comment]
**[Paul Benedict](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=ID134349)** commented

Any chance these negative durations occur during the switch to/from Daylight Savings Time?

**Comment by jira-importer** [comment]
**[Christian Jung](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=christian.jung)** commented

According to the documentation of `getCurrentTimeMillis()` it should be

```
the difference, measured in milliseconds, between the current time and midnight, January 1, 1970 UTC.
```

So it is affected by time adjustments (manual or NTP), but not by time zone changes and DST.
Of course some Java implementations could be buggy, so time zones and DST become visible as well.

**Comment by jira-importer** [comment]
**[Paul Benedict](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=ID134349)** commented

I know how it should work but I'd like to confirm this wasn't during DST transition, for the sake of ruling out the obvious.

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

Would anyone mind if I'd throw an `IllegalArgumentException` from the duration formatter if a negative duration has been passed? That would make the build fail, at least. After that, we need to find a way either tackle the Maven problem or prepare a workaround.

Again, [Christian Jung](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=christian.jung), are you able to write a single class file which reproduces the problem? I wasn't able to do so on Windows and FreeBSD.

**Comment by jira-importer** [comment]
**[Hervé Boutemy](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=hboutemy)** commented

IMHO, IllegalArgumentException (then failing the build) is too much: if negative, this means someone changed the system date at OS level, then the time measure is just inaccurate (and nothing can prevent it)
then just use 0
and eventually output a warning with the effective value measured, with this "time probably changed at OS level (manually or perhaps NTP time adjustement)" explanation

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

[Hervé Boutemy](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=hboutemy), makes sense because the formatter is not a general purpose library. But we could do the following:

1. Measure durations with `nanoTime`. The sources provided by Christian show that `nanoTime` is far more reliable than `currentTimeMillis`. The interfaces will remain the same but we will only divide by 10^6 because
2. `formatDuration` can return something like `INCORRECT` along with `log.warn`.

Probably something else.

**Comment by jira-importer** [comment]
**[Hervé Boutemy](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=hboutemy)** commented

+1
[from javadoc](http://docs.oracle.com/javase/7/docs/api/java/lang/System.html#nanoTime()), nanoTime() seems to have been created (in Java 5, so no problem) exactly for this use case

I don't know if my OS supports CLOCK_MONOTONIC, but using nanoTime() should really improve things: don't know if the warning will still happen, but it is a good defensive code

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

Hervé,

here are the spots:

```
maven-aether-provider/src/test/java/org/apache/maven/repository/internal/util/ConsoleTransferListener.java:            long duration = System.currentTimeMillis() - resource.getTransferStartTime();
maven-compat/src/main/java/org/apache/maven/artifact/repository/metadata/DefaultRepositoryMetadataManager.java:                    file.setLastModified( System.currentTimeMillis() );
maven-compat/src/main/java/org/apache/maven/repository/legacy/DefaultUpdateCheckManager.java:                props.setProperty( key, Long.toString( System.currentTimeMillis() ) );
maven-compat/src/main/java/org/apache/maven/repository/legacy/DefaultWagonManager.java:                                destination.setLastModified( System.currentTimeMillis() );
maven-compat/src/main/java/org/apache/maven/repository/legacy/MavenArtifact.java:        this.transferStartTime = System.currentTimeMillis();
maven-compat/src/test/java/org/apache/maven/artifact/testutils/TestFileManager.java:        File dir = new File( TEMP_DIR_PATH, baseFilename + System.currentTimeMillis() );
maven-compat/src/test/java/org/apache/maven/repository/legacy/StringWagon.java:            resource.setLastModified( System.currentTimeMillis() );
maven-core/src/main/java/org/apache/maven/artifact/repository/metadata/AbstractRepositoryMetadata.java:            metadataFile.setLastModified( System.currentTimeMillis() );
maven-core/src/main/java/org/apache/maven/lifecycle/internal/builder/BuilderCommon.java:        long buildEndTime = System.currentTimeMillis();
maven-core/src/main/java/org/apache/maven/lifecycle/internal/LifecycleModuleBuilder.java:        long buildStartTime = System.currentTimeMillis();
maven-core/src/main/java/org/apache/maven/lifecycle/internal/LifecycleModuleBuilder.java:            long buildEndTime = System.currentTimeMillis();
maven-core/src/test/java/org/apache/maven/project/DefaultMavenProjectBuilderTest.java:        localRepoDir = new File( System.getProperty( "java.io.tmpdir" ), "local-repo." + System.currentTimeMillis() );
maven-core/src/test/java/org/apache/maven/project/MavenProjectTest.java:            new File( System.getProperty( "java.io.tmpdir" ), "maven-project-tests" + System.currentTimeMillis()
maven-embedder/src/main/java/org/apache/maven/cli/event/ExecutionEventLogger.java:        long finish = System.currentTimeMillis();
maven-embedder/src/main/java/org/apache/maven/cli/transfer/AbstractMavenTransferListener.java:            long duration = System.currentTimeMillis() - resource.getTransferStartTime();
maven-embedder/src/main/java/org/apache/maven/cli/transfer/Slf4jMavenTransferListener.java:            long duration = System.currentTimeMillis() - resource.getTransferStartTime();
```

Did I miss some?

**Comment by jira-importer** [comment]
**[Hervé Boutemy](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=hboutemy)** commented

all occurrences of System.currentTimeMillis() are not causes of problems: last modified for repository, for example, is really requiring this API

the interesting cases for this issue seem

```
maven-core/src/main/java/org/apache/maven/lifecycle/internal/builder/BuilderCommon.java:        long buildEndTime = System.currentTimeMillis();
maven-core/src/main/java/org/apache/maven/lifecycle/internal/LifecycleModuleBuilder.java:        long buildStartTime = System.currentTimeMillis();
maven-core/src/main/java/org/apache/maven/lifecycle/internal/LifecycleModuleBuilder.java:            long buildEndTime = System.currentTimeMillis();
```

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

But what about

```
maven-aether-provider/src/test/java/org/apache/maven/repository/internal/util/ConsoleTransferListener.java:            long duration = System.currentTimeMillis() - resource.getTransferStartTime();
maven-compat/src/main/java/org/apache/maven/repository/legacy/MavenArtifact.java:        this.transferStartTime = System.currentTimeMillis();
maven-embedder/src/main/java/org/apache/maven/cli/event/ExecutionEventLogger.java:        long finish = System.currentTimeMillis();
maven-embedder/src/main/java/org/apache/maven/cli/transfer/AbstractMavenTransferListener.java:            long duration = System.currentTimeMillis() - resource.getTransferStartTime();
maven-embedder/src/main/java/org/apache/maven/cli/transfer/Slf4jMavenTransferListener.java:            long duration = System.currentTimeMillis() - resource.getTransferStartTime();
```

?

**Comment by jira-importer** [comment]
**[Hervé Boutemy](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=hboutemy)** commented

AFAIK, more about transfert, so:
- not really the actual issue,
- not so sensitive display (temporary during the download)
- takes some time, so should be less affected by little clock adjustments
- not so easy to be sure this code is the only one using stored value: resource.getTransferStartTime() can be used by other code

so IMHO, it is safer to let these parts as is for the moment and take more time to study

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

But at least this?

```
maven-embedder/src/main/java/org/apache/maven/cli/event/ExecutionEventLogger.java:        long finish = System.currentTimeMillis();
```

**Comment by jira-importer** [comment]
**[Christian Jung](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=christian.jung)** commented

[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o) Mabe its impossible to do that, because it would break out of the java sandbox (see e.g. [here](http://stackoverflow.com/questions/6203857/how-can-i-set-the-system-time-in-java))? In any case it requires admin privileges.

IMHO I'd like the negative durations displayed with a minus sign, plus maybe a warning with explanations on the log.
It's just about printing a minus sign and multiplying the duration with -1 if it's negative.

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

> IMHO I'd like the negative durations displayed with a minus sign, plus maybe a warning with explanations on the log. It's just about printing a minus sign and multiplying the duration with -1 if it's negative.

I do not think that this is that easy.

1. Just because the duration is negative we cannot assume that it is the same as the absolute value of the actual build time.
2. That negative time has been subtracted from the grant total session build time, so that value is broken. See your example.

I would try to fix what [Hervé Boutemy](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=hboutemy) proposes with a snapshot and then we will see whether it works or not.

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

Christian, I am planning to replace `System.currentTimeMillis` with `System.nanoTime` and truncate to milliseconds. Are you willing to test a branch?

**Comment by jira-importer** [comment]
**[Christian Jung](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=christian.jung)** commented

Michael, If this helps, I can repeat the test where i changed the system time with a batch script.

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

Great, I will provide a patched version next week.

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

Christian,

I am currently patching the source code and noticed some issues which out of our control. At [some points](https://github.com/eclipse/aether-core/search?p=1&q=currenttimemillis&utf8=%E2%9C%93) in Eclipse Aether `System.currentTimeMillis` is used for start times or durations. That needs to be changed to nanotime and scaled down to milliseconds. Please file an issues with Aether and link to this one.

Having a look at our code and the JavaDoc of `nanoTime` it does not look that simple because `nanoTime` is **not** related to the wall clock at all, it is arbitrary. It would be completely wrong to convert from `nanoTime` to `currentTimeMillis`. That would require a major change in Maven thus, affecting interfaces and users of Maven Embedder. This cannot happen before 3.3. I will raise another issue for that.

Read [this](http://stackoverflow.com/a/1776053/696632) SO answer by a Google employee.

**Comment by jira-importer** [comment]
**[Christian Jung](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=christian.jung)** commented

I filed a bug there: https://bugs.eclipse.org/bugs/show_bug.cgi?id=455944

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

Thanks, Aether isn't that crucial because it is used to measure the download/upload time only but still, there has to be a clear distinction between startTime as `Date` and startTime as `long` with nanosecond resolution. I have filed already MNG-5729 to address the issue in Maven. We should continue the discussion there. If that is fixed, this issue is fixed implicitly.

**Comment by jira-importer** [comment]
**[Kristian Rosenvold](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=krosenvold)** commented

Why doesn't vmware fix its crappy clock instead ? Did someone file an issue there ?

**Comment by jira-importer** [comment]
**[Christian Jung](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=christian.jung)** commented

`@Kristian` Rosenvold: I dont think this has to do with vmware. Its a question of wall clock and elapsed time. You cannot just subtract two wall clock times to get elapsed time.

**Comment by jira-importer** [comment]
**[Kristian Rosenvold](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=krosenvold)** commented

I'm sure there's lot of other minor cases of bad clocks too. I just felt the need to rant about the extraordinarily crappy clock inside vmware and other virtualisation solotions :) I did not intend to fix anything

**Comment by jira-importer** [comment]
**[Christian Schulte](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=schulte77)** commented

`System.currentTimeMillis()` is affected by everything changing the system clock. If e.g. a NTP client adjusts the system clock, negative durations can occur. The javadoc of class `java.util.time.Clock` contains further information (leap seconds, etc.). `System.nanoTime()` can only be used to calculate durations. It's more like a VM uptime counter in nanoseconds. There is no alternative to using `System.currentTimeMillis()` or `java.util.time.Clock.systemUTC()` whenever a real timestamp is needed.

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

No one's talking about a replacement for a timestamp but merely not to calculate elapsed time based on `System.currentTimeMillis()` but only on `System.nanoTime()`.

**Comment by jira-importer** [comment]
**[Christian Schulte](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=schulte77)** commented

Ok. Are you planning to work on this?

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

I made some initial analysis on this more than year ago (MNG-5729). I would involve to change a lot of spots. This should not happen before 3.4 or even 4.0 and how critical is this? Given that Maven is not a long-running process and this has been reported only once, I'd say this is low priority. We have other important issues. I am currently loaded with Doxia Sitetools 1.7 preparations anyway.

If you have some spare time, give it a try in a branch 'MNG-5729'. More interestingly, how can we create a proper IT for that?

**Comment by jira-importer** [comment]
**[Hervé Boutemy](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=hboutemy)** commented

+1
not sure an IT is so much important, since I fear it will be really hard to create, but given the focused changed, I would trust the fix from source update: what is important is to be sure it won't wreck havoc on plugins, extensions, or anything outside core

**Comment by jira-importer** [comment]
**[Guillaume Nodet](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=gnt)** commented

It will be fixed as part of MNG-5729
